src/server/qgsserver.cpp

```diff
@@ -344,33 +355,54 @@ void QgsServer::handleRequest( QgsServerRequest &request, QgsServerResponse &res
       const QgsServerParameters params = request.serverParameters();
       printRequestParameters( params.toMap(), logLevel );
 
-      //Config file path
+      // Setup project (config file path)
       if ( ! project )
       {
+        QString configFilePath = configPath( *sConfigFilePath, params.map() );
+
         // load the project if needed and not empty
-        project = mConfigCache->project( sServerInterface->configFilePath() );
-        if ( ! project )
-        {
-          throw QgsServerException( QStringLiteral( "Project file error" ) );
-        }
+        project = mConfigCache->project( configFilePath );
       }
 
-      if ( ! params.fileName().isEmpty() )
+      if ( project )
       {
-        const QString value = QString( "attachment; filename=\"%1\"" ).arg( params.fileName() );
-        requestHandler.setResponseHeader( QStringLiteral( "Content-Disposition" ), value );
+        sServerInterface->setConfigFilePath( project->fileName() );
       }
 
-      // Lookup for service
-      QgsService *service = sServiceRegistry->getService( params.service(), params.version() );
-      if ( service )
+      // Dispatcher: if SERVICE is set, we assume a OWS service, if not, let's try an API
+      // TODO: QGIS 4 fix the OWS services and treat them as APIs
+      QgsServerApi *api = nullptr;
+      if ( params.service().isEmpty() && ( api = sServiceRegistry->getApiForRequest( request ) ) )
       {
-        service->executeRequest( request, responseDecorator, project );
+        QgsServerApiContext context { api->rootPath(), &request, &response, project, sServerInterface };
```

Comment:
You pass the raw response object to the api and not the `responseDecorator`. The goal of the decorator is to encapsulate the call of plugin's filter's hook.  As far as I understand, if the `QgsServerApiContext` does not handle it in some way, then using api call will disable the use of plugins.

Reply:
Yeah, I did not yet address the plugins integration, the ACL plugins are also in my TODO list.

Reply:
@dmarteau does https://github.com/qgis/QGIS/pull/10016/commits/5893f814ba334c0d112fcd89e2fd57a3a2c58b66 address this issue?
I had to expose the decorator to Python mainly for testability, if you think it's an issue I can reimplement the API registering test in C++, but I think it could be useful to be able to construct and test the API context in Python.

Reply:
@elpaso You don't have to (and AMHA you shouldn't) declare  QgsFilterResponseDecorator explicitely in the QgsServerApiContext constructor, because QgsFilterResponseDecorator inherits from QgsServerResponse. 

The signature was fine, it's just when you call  the QgsServerApiContext constructor you have to pass a QgsFilterResponseDecorator - as for service->executeRequest(): it should makes no différence to your code.

Reply:
Sorry, didn't fully read your comment ....

Anyway, I think there is potential issue here because you force usage of an internal object and we should rely on polymorphism. Also, I do not fully understand why you need to explicitely declare a QgsFilterResponseDecorator, because SIP supports polymorphism.

Reply:
You are right, passing the decorator into the context instead of the response seems the easier solution